sim/prng.js

```diff
@@ -96,6 +96,24 @@ class PRNG {
 		return this.next(denominator) < numerator;
 	}
 
+	/**
+	 * @nocommit
+	 *
+	 * @param {ReadonlyArray<T>} items
+	 * @return {T}
+	 * @template T
+	 */
+	randomItem(items) {
+		if (items.length === 0) {
+			throw new RangeError(`Cannot sample an empty array`);
+		}
+		const index = this.next(items.length);
+		if (!(index in items)) {
```

Comment:
The ``in`` check doesn't look great performance-wise for the common case of compact arrays.
In fact, we shouldn't have any sparse arrays in PS code at all...

Reply:
Yeah, PS doesn't use sparse arrays at all; we have objects and Maps for that.

Reply:
The check is designed to catch mistakes.

I'll measure and see if this check is a performance problem.

Reply:
I measured. `in` does have a time cost. However, I found a method which *doesn't* have a cost.

Microbenchmarks: https://jsperf.com/pokemon-showdown-random-item-sparse-array-check
Results on my laptop (Intel(R) Core(TM) i7-3720QM CPU @ 2.60GHz): https://imgur.com/qPhrQeo
Results on my desktop (AMD Phenom(tm) II X6 1055T Processor): https://imgur.com/uWyQejK

Reply:
I'm torn on this. It seems like such a random(*) place for an assertion... The method will sometimes work with sparse arrays and sometimes won't, and throwing <em>sometimes</em> with the same input looks... like a recipe for a headache. Therefore, while this intends to help with development, it might be counterproductive.

(*) Pun intended.

Reply:
> The method will sometimes work with sparse arrays and sometimes won't, and throwing *sometimes* with the same input looks... like a recipe for a headache.

I agree.

My hypothesis is that the deterministic nature of `PRNG` will make errors reproducible, so errors thrown by `randomItem` will be easy to debug. This might not be the case for the team generator though; I have no experience using that code.

> Therefore, while this intends to help with development, it might be counterproductive.

Based on the current uses of `random`/`next` for an array index, here's what will happen to the return value of `randomItem`:

* Use as the object in method call: `randomItem(...).id`
* Use in a string context: `"hello " + randomItem(...)`
* Use in a numeric context: `3 * randomItem(...)`

Let's say we make `randomItem` return `undefined` (i.e. no check) instead of throwing an exception in the case that `items` is a parse array and the selected index is not present in the array.

In the first case (method call), an exception will be thrown anyway, but possibly far away from the call to `randomItem` or the construction of the input array.

In the second case (string), someone will see a string containing `undefined` somewhere. No exception will be thrown, so it might be hard to find where the `'undefined'` string is created, let alone where the `randomItem` call is.

In the third case (number), in the best case, someone will see `NaN` somewhere. No exception will be thrown, and since `NaN` propagates, it will be hard to find where the `NaN` originated from, let alone where the `randomItem` call is. (Maybe there are debugger options to find `NaN`? I'm not sure.)

Therefore, I think throwing an exception is better than returning `undefined`.

> Can't this problem be deferred to TypeScript somehow?

I think TypeScript doesn't and can't distinguish sparse arrays from dense arrays.

The only way I think TypeScript can help if we type `randomItem` to return `T | undefined`. This would put a huge burden on callers; partically every user of `randomItem` would need to check for `undefined` to make TypeScript happy. I think that problem would overshadow the advantage of having a `randomItem` function in the first place.

Reply:
> My hypothesis is that the deterministic nature of PRNG will make errors reproducible, so errors thrown by randomItem will be easy to debug. This might not be the case for the team generator though; I have no experience using that code.

It would actually be easy to deal with team generators. However, I'd venture to say that only a very small minority of PS devs actually know how to debug them...

In the end, your arguments are not weak, but I am still -0 on this check... I guess I'll let Zarel decide, since he's more knowledgeable about TypeScript capabilities.

Reply:
> I am still -0 on this check

What do you mean by "-0"?